dot/parachain/candidate-validation/candidate_validation_test.go

```diff
@@ -677,3 +669,190 @@ func Test_performBasicChecks(t *testing.T) {
 		})
 	}
 }
+
+func TestCandidateValidation_validateFromChainState(t *testing.T) {
+	t.Parallel()
+	candidateReceipt, validationCode := createTestCandidateReceiptAndValidationCode(t)
+	candidateReceipt2 := candidateReceipt
+	candidateReceipt2.Descriptor.PovHash = common.MustHexToHash(
+		"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef")
+	candidateReceipt2.Descriptor.ParaID = 2
+
+	candidateReceipt3 := candidateReceipt
+	candidateReceipt3.Descriptor.ParaID = 3
+
+	candidateReceipt4 := candidateReceipt
+	candidateReceipt4.Descriptor.ParaID = 4
+	candidateReceipt4.Descriptor.ValidationCodeHash = parachaintypes.ValidationCodeHash(common.MustHexToHash(
+		"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"))
+
+	candidateReceipt5 := candidateReceipt
+	candidateReceipt5.Descriptor.ParaID = 5
+
+	povHashMismatch := PoVHashMismatch
+	paramsTooLarge := ParamsTooLarge
+	codeHashMismatch := CodeHashMismatch
+	badSignature := BadSignature
+
+	ctrl := gomock.NewController(t)
+	t.Cleanup(ctrl.Finish)
+
+	// NOTE: adder parachain internally compares postState with bd.State in it's validate_block,
+	// so following is necessary.
+	encodedState, err := scale.Marshal(uint64(1))
+	require.NoError(t, err)
+	postState, err := common.Keccak256(encodedState)
+	require.NoError(t, err)
+
+	hd, err := scale.Marshal(HeadDataInAdderParachain{
+		Number:     uint64(1),
+		ParentHash: common.MustHexToHash("0x0102030405060708090001020304050607080900010203040506070809000102"),
+		PostState:  postState,
+	})
+	require.NoError(t, err)
+
+	expectedPersistedValidationData := parachaintypes.PersistedValidationData{
+		ParentHead:             parachaintypes.HeadData{Data: hd},
+		RelayParentNumber:      uint32(1),
+		RelayParentStorageRoot: common.MustHexToHash("0x50c969706800c0e9c3c4565dc2babb25e4a73d1db0dee1bcf7745535a32e7ca1"),
+		MaxPovSize:             uint32(2048),
+	}
+
+	expectedPersistedValidationDataSmallMax := parachaintypes.PersistedValidationData{
+		ParentHead:             parachaintypes.HeadData{Data: hd},
+		RelayParentNumber:      uint32(1),
+		RelayParentStorageRoot: common.MustHexToHash("0x50c969706800c0e9c3c4565dc2babb25e4a73d1db0dee1bcf7745535a32e7ca1"),
+		MaxPovSize:             uint32(10),
+	}
+
+	mockInstance := NewMockRuntimeInstance(ctrl)
+	mockInstance.EXPECT().
+		ParachainHostPersistedValidationData(
+			uint32(1000),
+			gomock.AssignableToTypeOf(parachaintypes.OccupiedCoreAssumption{})).
+		Return(&expectedPersistedValidationData, nil).Times(1)
```

Comment:
If we expect to call the function only once, It is not required to write `.Times(1)`.